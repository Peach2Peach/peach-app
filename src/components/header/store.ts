import create from 'zustand'

type HeaderConfig = {
  title?: string
  titleComponent?: JSX.Element
  icons?: { iconComponent: JSX.Element; onPress: () => void }[]
  showGoBackButton?: boolean
}

type HeaderState = HeaderConfig & {
  setHeaderState: (headerConfiguration: HeaderConfig) => void
}

const defaultState = {
  title: '',
  titleComponent: undefined,
  icons: [],
  showGoBackButton: true,
}

export const useHeaderState = create<HeaderState>()((set) => ({
  ...defaultState,
  setHeaderState: (props) => set({ ...defaultState, ...props }),
}))
