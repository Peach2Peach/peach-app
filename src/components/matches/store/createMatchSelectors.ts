import { getAvailableCurrencies } from '../../../utils/match'
import { hasMoPsInCommon, getMoPsInCommon, getPaymentMethods } from '../../../utils/paymentMethod'

export type MatchSelectors = {
  [id: Match['offerId']]: {
    selectedCurrency?: Currency
    selectedPaymentMethod?: PaymentMethod
    availableCurrencies: Currency[]
    availablePaymentMethods: PaymentMethod[]
    mopsInCommon: MeansOfPayment
    meansOfPayment: MeansOfPayment
    showCurrencyPulse: boolean
    showPaymentMethodPulse: boolean
  }
}

export const createMatchSelectors = (matches: Match[], offerMeansOfPayment: MeansOfPayment) =>
  matches.reduce((acc: MatchSelectors, match) => {
    const mopsInCommon = hasMoPsInCommon(offerMeansOfPayment, match.meansOfPayment)
      ? getMoPsInCommon(offerMeansOfPayment, match.meansOfPayment)
      : match.meansOfPayment

    const availableCurrencies = getAvailableCurrencies(mopsInCommon, match.meansOfPayment)
    const availablePaymentMethods = getPaymentMethods(mopsInCommon)

    acc[match.offerId] = {
      selectedCurrency: availableCurrencies.length === 1 ? availableCurrencies[0] : undefined,
      selectedPaymentMethod: availablePaymentMethods.length === 1 ? availablePaymentMethods[0] : undefined,
      availableCurrencies,
      availablePaymentMethods,
      mopsInCommon,
      meansOfPayment: match.meansOfPayment,
      showCurrencyPulse: false,
      showPaymentMethodPulse: false,
    }
    return acc
  }, {})
