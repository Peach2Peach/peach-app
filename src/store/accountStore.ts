import { create } from 'zustand'
import { createJSONStorage, persist } from 'zustand/middleware'
import { createStorage, toZustandStorage } from '../utils/storage'

type Account = {
  identity?: Identity
  loggedIn: boolean
  migrated: boolean
}
type AccountStore = Account & {
  setIdentity: (identity: Identity) => void
  setMigrated: () => void
  reset: () => void
}

const defaultAccount: Account = {
  identity: undefined,
  loggedIn: false,
  migrated: false,
}

export const accountStorage = createStorage('accountStorage')

export const useAccountStore = create(
  persist<AccountStore>(
    (set) => ({
      ...defaultAccount,
      setIdentity: (identity) => set({ identity, loggedIn: !!identity.privKey }),
      setMigrated: () => set({ migrated: true }),
      reset: () => set(defaultAccount),
    }),
    {
      name: 'account',
      version: 0,
      storage: createJSONStorage(() => toZustandStorage(accountStorage)),
    },
  ),
)
