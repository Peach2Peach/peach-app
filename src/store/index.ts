import create from 'zustand'
import { immer } from 'zustand/middleware/immer'
import { info } from '../utils/log'
import { AccountStore, createAccountSlice } from './createAccountSlice'
import { ContractsStore, createContractsSlice } from './createContractsSlice'
import { createOffersSlice, OffersStore } from './createOffersSlice'
import { createPaymentDataSlice, PaymentDataStorage } from './createPaymentDataSlice'

type UserDataFunctions = { initialize: () => void }

export type UserDataStore = AccountStore & ContractsStore & OffersStore & PaymentDataStorage & UserDataFunctions

export const useUserDataStore = create<UserDataStore>()(
  immer((...a) => ({
    ...createAccountSlice(...a),
    ...createContractsSlice(...a),
    ...createOffersSlice(...a),
    ...createPaymentDataSlice(...a),
    initialize: () => {
      info('useUserDataStore - initialize - start')
      a[1]().initializeContracts()
      a[1]().initializeOffers()
      a[1]().initializePaymentData()
      info('useUserDataStore - initialize - done')
    },
  })),
)
