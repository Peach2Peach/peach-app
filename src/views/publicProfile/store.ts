import create from 'zustand'

type ProfileStore = {
  currentUserPubkey: string
  setCurrentUserPubkey: (currentUserPubkey: string) => void
}

export const useProfileStore = create<ProfileStore>()((set) => ({
  currentUserPubkey: '',
  setCurrentUserPubkey: (currentUserPubkey: string) => set({ currentUserPubkey }),
}))
