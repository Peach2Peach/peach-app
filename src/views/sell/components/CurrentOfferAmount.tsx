import { BTCAmount } from '../../../components/bitcoin'
import { useOfferPreferences } from '../../../store/offerPreferenes'

export const CurrentOfferAmount = () => {
  const amount = useOfferPreferences((state) => state.sellAmount)

  return <BTCAmount size="small" amount={amount} />
}
