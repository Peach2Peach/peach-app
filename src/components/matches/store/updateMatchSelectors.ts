import { MatchSelectors } from './createMatchSelectors'

export const updateMatchSelectors = (matchSelectors: MatchSelectors, newMatchSelectors: MatchSelectors) => {
  const updatedMatchSelectors = Object.keys(matchSelectors).reduce((acc: MatchSelectors, matchId) => {
    if (newMatchSelectors[matchId]) {
      acc[matchId] = {
        ...matchSelectors[matchId],
        meansOfPayment: newMatchSelectors[matchId].meansOfPayment,
      }
    }
    return acc
  }, {})
  return updatedMatchSelectors
}
