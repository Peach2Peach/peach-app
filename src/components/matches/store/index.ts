import { create } from 'zustand'
import { immer } from 'zustand/middleware/immer'
import { getAvailableMethods } from '../../../utils/match'
import { createMatchSelectors, MatchSelectors } from './createMatchSelectors'
import { updateMatchSelectors } from './updateMatchSelectors'

type MatchState = {
  matchSelectors: MatchSelectors
  currentIndex: number
  currentPage: number
}

export type MatchStore = MatchState & {
  setSelectedCurrency: (currency: Currency, matchId: Match['offerId']) => void
  setSelectedPaymentMethod: (paymentMethod: PaymentMethod | undefined, matchId: Match['offerId']) => void
  setCurrentIndex: (newIndex: number) => void
  setAvailablePaymentMethods: (methods: PaymentMethod[], matchId: Match['offerId']) => void
  resetStore: () => void
  addMatchSelectors: (matches: Match[], offerMeansOfPayment: MeansOfPayment) => void
  setShowMissingPaymentMethodWarning: (matchId: Match['offerId'], show?: boolean) => void
}

const defaultState: MatchState = {
  currentIndex: 0,
  currentPage: 0,
  matchSelectors: {},
}

export const useMatchStore = create<MatchStore>()(
  immer((set, get) => ({
    ...defaultState,
    setSelectedCurrency: (currency, matchId) => {
      const currentMatch = get().matchSelectors[matchId]
      const currentPaymentMethod = currentMatch.selectedPaymentMethod
      const newMethods = getAvailableMethods(currentMatch.meansOfPayment, currency, currentMatch.mopsInCommon)
      get().setAvailablePaymentMethods(newMethods, matchId)
      if (newMethods.length === 1) {
        get().setSelectedPaymentMethod(newMethods[0], matchId)
      } else if (currentPaymentMethod && !newMethods.includes(currentPaymentMethod)) {
        get().setSelectedPaymentMethod(undefined, matchId)
      }
      return set((state) => {
        state.matchSelectors[matchId].selectedCurrency = currency
      })
    },
    setSelectedPaymentMethod: (paymentMethod, matchId) =>
      set((state) => {
        state.matchSelectors[matchId].showMissingPaymentMethodWarning = false
        state.matchSelectors[matchId].selectedPaymentMethod = paymentMethod
      }),
    setCurrentIndex: (newIndex) =>
      set((state) => ({ ...state, currentIndex: newIndex, currentPage: Math.floor(newIndex / 10) })),
    setAvailablePaymentMethods: (methods, matchId) =>
      set((state) => {
        state.matchSelectors[matchId].availablePaymentMethods = methods
      }),
    resetStore: () => set(() => defaultState),
    addMatchSelectors: (matches, offerMeansOfPayment) => {
      const newMatchSelectors = createMatchSelectors(matches, offerMeansOfPayment)
      const updatedMatchSelectors = updateMatchSelectors(get().matchSelectors, newMatchSelectors)

      return set((state) => ({
        ...state,
        matchSelectors: { ...newMatchSelectors, ...updatedMatchSelectors },
      }))
    },
    setShowMissingPaymentMethodWarning: (matchId, show = true) =>
      set((state) => {
        state.matchSelectors[matchId].showMissingPaymentMethodWarning = show
      }),
  })),
)
