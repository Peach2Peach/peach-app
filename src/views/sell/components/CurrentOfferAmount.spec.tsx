import { CurrentOfferAmount } from './CurrentOfferAmount'
import { createRenderer } from 'react-test-renderer/shallow'
import { useOfferPreferences } from '../../../store/offerPreferenes'

describe('CurrentOfferAmount', () => {
  it('should render correctly', () => {
    const renderer = createRenderer()
    useOfferPreferences.setState({ sellAmount: 21000000 })
    renderer.render(<CurrentOfferAmount />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
