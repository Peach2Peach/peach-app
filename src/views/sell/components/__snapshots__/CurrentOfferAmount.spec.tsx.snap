// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentOfferAmount should render correctly 1`] = `
<BTCAmount
  amount={21000000}
  size="small"
/>
`;
